doc/developer/design/19042023_stash_migrations.md

```diff
@@ -0,0 +1,352 @@
+    that all of our keys and values must implement. A requirement of `Data` is that the type must
+    implement [`Ord`](https://doc.rust-lang.org/std/cmp/trait.Ord.html), which seems to be required
+    for consolidating rows. `serde_json::Value` is not order-able, and there isn't a good way to
+    define how a generic JSON value should be ordered.
+* Maintaining `objects_vX.rs` with Rust types instead of an `objects_vX.proto`.
+    * Instead of defining our Stash types in protobuf we could keep them defined in Rust and follow
+    the same approach for snapshotting the `objects_vX.rs` files these types would be in. 
+    * ❌ An issue with this approach is "snapshot hygiene". Nothing would prevent us from adding 
+    dependencies to these snapshots, and to be correct we'd also need to also snapshot the
+    dependency. But we don't have a tool to assert dependencies are snapshotted, so a dependency
+    could change which could silently introduce a change to the serialized format, causing bugs.
+
+### Why this approach?
+
+I believe this is the right design because of the following reasons:
+
+* It allows us to write easy to reason about migration functions that convert from `TypeA` to `TypeB`.
+* Introduces automated checks to make sure a migration exists whenever Stash objects change.
+* Old migrations do not need to change when new fields are introduced, since they can operate on
+old versions of the types.
+* Introduces the possibility of simulation testing for upgrades, since we maintain the history of
+the schema of the Stash, through snapshotting.
+
+Overall I believe it achieves the goal of "fearless migrations", for the lowest possible
+engineering cost.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* How important do we believe human readability of the Stash is?
+* Does the Stash need to be JSON for billing purposes? This [issue](https://github.com/MaterializeInc/materialize/issues/14264)
+  indicates it might, but this [comment](https://github.com/MaterializeInc/materialize/issues/14264#issuecomment-1218250985)
+  indicates it does not.
+* Is it important to maintain uniqueness checks when doing Stash migrations?
+
+# Future work
+[future-work]: #future-work
+
+* As mentioned in [testing-and-observability] the most immediate followup work could be to
+generate old versions of the Stash using our snapshotted types for testing. We could also generate
+very large Stashes for load testing.
+* Not necessarily related to migrations, but if the Stash is not human readable, and requires
+`stash-debug`, that opens up the possibility of compressing the SQL statements stored in the
+`create_sql` column of the `"items"` collection with an algorithm like brotli. Theoretically
+this reduces the amount of data we're storing in CockroachDB, and improves the speed of the Stash.
```

Comment:
FWIW Cockroach uses Snappy to compress data before it hits the disk, so additional compression likely won't move the needle on how much disk storage is required. It could in theory reduce the amount of data we're transiting to Cockroach over pgwire but .. I think a massive `CREATE VIEW` statement is out of spec for Materialize. I'd be willing to bet that folks are using massive `CREATE VIEW`s with static data only because our table support is limited, and big `VALUES` clauses are a way to work around that.

Reply:
That totally makes sense, I'll call this out in the doc, thanks!